pyro/poutine/trace.py

```diff
@@ -127,93 +286,58 @@ def log_pdf(self, site_filter=lambda name, site: True):
         :returns: total log probability.
         :rtype: torch.autograd.Variable
         """
-        log_p = 0.0
-        for name, site in self.nodes.items():
-            if site["type"] == "sample" and site_filter(name, site):
-                try:
-                    site_log_p = site["log_pdf"]
-                except KeyError:
-                    args, kwargs = site["args"], site["kwargs"]
-                    site_log_p = site["fn"].log_prob(site["value"], *args, **kwargs)
-                    site_log_p = scale_tensor(site_log_p, site["scale"]).sum()
-                    site["log_pdf"] = site_log_p
-                    _warn_if_nan(name, site_log_p)
-                log_p += site_log_p
-        return log_p
-
-    def compute_batch_log_pdf(self, site_filter=lambda name, site: True):
+        return fold_trace(lambda log_p, msg: log_p + log_pdf(msg),
```

Comment:
As a lover of lambda calculus, I still have to :-1: this use of lambdas in Python. A for loop would be clearer than `fold`.